File: qqbot-persistence/src/main/java/com/mieai/qqbot/persistence/internal/PersistenceValidation.java
package com.mieai.qqbot.persistence.internal;

import java.util.Objects;
import java.util.UUID;

/** Shared validation for persistence input values that become keys or discriminator columns. */
public final class PersistenceValidation {
    private static final UUID NIL_UUID = new UUID(0L, 0L);

    private PersistenceValidation() {
    }

    public static UUID requireIdentifier(UUID value, String name) {
        Objects.requireNonNull(value, name + " must not be null");
        if (NIL_UUID.equals(value)) {
            throw new IllegalArgumentException(name + " must not be the nil UUID");
        }
        return value;
    }

    public static String requireToken(String value, String name) {
        Objects.requireNonNull(value, name + " must not be null");
        if (value.isBlank()) {
            throw new IllegalArgumentException(name + " must not be blank");
        }
        if (!value.equals(value.strip())) {
            throw new IllegalArgumentException(name + " must not have surrounding whitespace");
        }
        if (value.codePoints().anyMatch(Character::isWhitespace)) {
            throw new IllegalArgumentException(name + " must not contain whitespace");
        }
        if (value.codePoints().anyMatch(Character::isISOControl)) {
            throw new IllegalArgumentException(name + " must not contain control characters");
        }
        return value;
    }

    public static String requireToken(String value, String name, int maxCharacters) {
        String normalized = requireToken(value, name);
        if (normalized.codePointCount(0, normalized.length()) > maxCharacters) {
            throw new IllegalArgumentException(
                    name + " must not exceed " + maxCharacters + " characters");
        }
        return normalized;
    }

    public static String requirePayload(String value, String name) {
        Objects.requireNonNull(value, name + " must not be null");
        if (value.isBlank()) {
            throw new IllegalArgumentException(name + " must not be blank");
        }
        return value;
    }
}

File: qqbot-persistence/src/main/kotlin/com/mieai/qqbot/persistence/internal/PersistenceValidation.kt
package com.mieai.qqbot.persistence.internal

import java.util.UUID

/** Shared validation for persistence input values that become keys or discriminator columns. */
object PersistenceValidation {
    private val nilUuid = UUID(0L, 0L)

    fun requireIdentifier(value: UUID, name: String): UUID {
        require(value != nilUuid) { "$name must not be the nil UUID" }
        return value
    }

    fun requireToken(value: String, name: String): String {
        require(value.isNotBlank()) { "$name must not be blank" }
        require(value == value.trim()) { "$name must not have surrounding whitespace" }
        require(value.codePoints().noneMatch { Character.isWhitespace(it) }) { "$name must not contain whitespace" }
        require(value.codePoints().noneMatch { Character.isISOControl(it) }) {
            "$name must not contain control characters"
        }
        return value
    }

    fun requireToken(value: String, name: String, maxCharacters: Int): String {
        val normalized = requireToken(value, name)
        require(normalized.codePointCount(0, normalized.length) <= maxCharacters) {
            "$name must not exceed $maxCharacters characters"
        }
        return normalized
    }

    fun requirePayload(value: String, name: String): String {
        require(value.isNotBlank()) { "$name must not be blank" }
        return value
    }
}
